src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs

```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                    Assert.True(false, "RemoteCertificateValidationCallback called when AuthenticateAsServerAsync was expected to fail.");
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    Assert.Equal(hostName, actualHostName);
+                    return serverCert;
+                };
+
+                var cts = new CancellationTokenSource();
+                Assert.Throws<InvalidOperationException>(() => {server.AuthenticateAsServerAsync(options, cts.Token).Wait();});
+
+                Assert.Equal(0, timesCallbackCalled);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackNotSet_UsesLocalCertificateSelection(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.Equal(targetHost, string.Empty);
+                Assert.True(localCertificates.Contains(serverCert));
+                timesCallbackCalled++;
+                return serverCert;
+            });
+
+            var validationCallback = new RemoteCertificateValidationCallback((object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+            {
+                Assert.Equal(serverCert, certificate);
+                return true;
+            });
+
+            VirtualNetwork vn = new VirtualNetwork();
+            using (VirtualNetworkStream serverStream = new VirtualNetworkStream(vn, isServer: true),
+                                        clientStream = new VirtualNetworkStream(vn, isServer: false))
+            using (SslStream server = new SslStream(serverStream, false, null, selectionCallback),
+                             client = new SslStream(clientStream, leaveInnerStreamOpen: false, validationCallback))
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificate = serverCert;
+
+                var cts = new CancellationTokenSource();
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
```

Comment:
```C#
await clientJob;
```

Reply:
And, actually, since the two awaits are related (such that if one failed the other might hang), it'd be better to do:
```C#
await TaskTimeoutExtensions.WhenAllOrAnyFailed(new[] { clientJob, server.AuthenticateAsServerAsync(options, CancellationToken.None) });
```